File: 프로그래머스/2/42587. 프로세스/프로세스.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(vector<int> priorities, int location) {
    vector<int>priorityCount(10, 0);
    vector<bool>complete(priorities.size(), false);
    for(int priority : priorities){
        priorityCount[priority]++;
    }
    
    int idx = 0;
    int processCount = 0;
    while(1){
        if(idx >= priorities.size()) idx = 0;
        if(complete[idx]) {
            idx++;
            continue;
        }
        int tmpPriority = priorities[idx];
        bool findBiggerPriority = false;
        for(int p = 9; p > tmpPriority; p--){
            if(priorityCount[p]){
                idx++;
                findBiggerPriority = true;
                break;
            }
        }
        if(findBiggerPriority) continue;
        
        complete[idx] = true;
        priorityCount[tmpPriority]--;
        processCount++;
        if(idx == location) return processCount;
    }
    return -1;
}

File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;

class Solution {
    static class Process {
        int location;
        int priority;
        
        public Process(int location, int priority) {
            this.location = location;
            this.priority = priority;
        }
    }
    
    public int solution(int[] priorities, int location) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        Queue<Process> q = new ArrayDeque<>();
        
        for(int i = 0; i < priorities.length; i++) {
            pq.add(priorities[i]);
            q.add(new Process(i, priorities[i]));
        }
        
        int curP = pq.poll();
        int answer = 1;
        while(true) {
            Process next = q.poll();
            if(next.priority == curP) {
                if(next.location == location) return answer;
                answer++;
                curP = pq.poll();
                continue;
            }
            q.add(next);
        }
    }
}